File: 671-2/D.py
n = int(input())
spheres = sorted(list(map(int, input().split())))
shop = [0 for i in range(n)]
l = 0

while len(spheres) > 0:
    top = spheres[-1]
    i = -2
    while -i <= len(spheres):
        if spheres[i] != top:
            break
        i -= 1
    i += 1
    spheres = spheres[:i]

    while i < 0 and l < n:
        shop[l] = top
        l += 2
        i += 1
    
    if l >= n:
        spheres += [top] * -i
        break

l = 1
leftover = []
cnt = 0
while len(spheres) > 0:
    c = spheres.pop()
    if l < n-1 and shop[l-1] > c and shop[l+1] > c and shop[l] == 0:
        shop[l] = c
        l += 2
        cnt += 1
    else:
        leftover.append(c)

for i in range(n):
    if shop[i] == 0:
        shop[i] = leftover.pop()

print(cnt)
print(*shop)

File: 671-2/D.cpp
#include <vector>
#include <cstdio>
#include <algorithm>
using namespace std;

int n, a, l, cnt;
vector<int> spheres, shop, leftover;

int main() {
    scanf("%d", &n);
    for (int i = 0; i < n; ++i) {
        scanf("%d", &a);
        spheres.emplace_back(a);
    }
    shop = vector<int>(n);
    sort(spheres.begin(), spheres.end());

    l = 0;
    while(spheres.size() > 0) {
        int top = spheres.back();
        spheres.pop_back();
        int i = 1;
        while (spheres.back() == top) {
            ++i;
            spheres.pop_back();
        }

        while (i > 0 && l < n) {
            shop[l] = top;
            l += 2;
            --i;
        }

        if (l >= n) {
            while (i > 0) {
                spheres.emplace_back(top);
                --i;
            }
            break;
        }
    }

    l = 1;
    while (spheres.size() > 0) {
        int c = spheres.back();
        spheres.pop_back();
        if (l < n-1 && shop[l-1] > c && shop[l+1] > c && shop[l] == 0) {
            shop[l] = c;
            l += 2;
        } else {
            leftover.emplace_back(c);
        }
    }

    for (int i = 0; i < n; ++i) {
        if (shop[i] == 0) {
            int c = leftover.back();
            leftover.pop_back();
            shop[i] = c;
        }
    }

    cnt = 0;
    for (int i = 1; i < n-1; ++i) {
        if (shop[i-1] > shop[i] && shop[i+1] > shop[i]) {
            ++cnt;
        }
    }

    printf("%d\n", cnt);
    for (int i = 0; i < n; ++i) {
        printf("%d ", shop[i]);
    }
}